Full refactor and update of labscript driver

dihm opened the pull request:
Finished the refactor of the labscript driver to match advice from @philipstarkey. To the surprise of no one, actually conforming to the design of labscript has led to significantly cleaner device code.

Basic structure is
- `PrawnDODevice`: a `PseudoclockDevice` that does all of the actual work.
  - `PrawnDoDevice.outputs`: an internal `IntermediateDevice` that holds all of the `DigitalOut` outputs. Outputs must be directly connected to it, not the `PrawnDODevice` to work.
- `PrawnDO`: a dummy `IntermediateDevice` that allows for directly connecting a PrawnDO to a Clockline (rather than via a DigitalOutput Trigger).
  - When using this device, the name supplied by the user is applied to the auto-created `PrawnDODevice`. So the user is actually interacting with that, not the dummy class.

Have confirmed that this code properly accounts for specified re-trigger delays. I've also tidied up my test rig (ie used cables all of the same length) and confirm the delay is 50ns.

Before merging this, should go ahead and merge #3. I'd also like to stress-test using the driver a bit more to make sure there aren't any edge cases that need to be fixed.

dihm committed:
Initial (untested) refactor to pseudoclock device

dihm committed:
Make PrawnDO properly pass DigitalOuts to internal PrawnDODevice
Minor fixes to error strings.

dihm committed:
Refactor now compiles.

Note: have not confirmed that retrigger offset times actually work
(which was the whole point of all this).
Also, this method produces two device tabs currently which needs to be fixed.

dihm committed:
Fix doubled BLACS tab, pass through user's name for device
when using IntermediateDevice.

dihm committed:
Fix issue with retrigger delays not being accounted for.

Also renamed a bunch of stuff under the hood.

dihm committed:
Add some checks to ensure specified channel names are valid.

Channel names can be much more varied as well, just have to end with a hex digit.

dihm committed:
Make channel labels match allowed channels.

philipstarkey commented:
Another possible tweak here could be to allow either a ClockLine or the standard digital trigger params to be provided to the PrawnDODevice class (as keyword arguments) and internally construct the PrawnDO device before calling the base class __init__, when needed. That way you don't have the weirdness with the name being passed to the child device thing.

My main reason for suggesting this is that if we ever get around to implementing the Python descriptor stuff that was discussed on the mailing list 6 months ago, then I think this would break, because that upcoming functionality will rely on the instance being assigned to the name also being the object that is instantiated by the user.

dihm committed:
Auto create Intermediate device as needed so there is only one
class that is necessary.

dihm commented:
That is a good call. That intermediate device isn't doing anything anyway, so it should get abstracted away.

Behavior is now that you specify `clock_line` to directly attach to a clockline or it uses `trigger_device` and `trigger_connection`. 

Also realized that I need to update the runviewer parser. Once that is done, hopefully this is ready to go.

dihm committed:
Update runviewer parser to actually load.

dihm commented:
Actually, the runviewer parser is turning out to be a bit more subtle than I thought. When using the dummy intermediate device to attach it to a clockling, runviewer tries to load the dummy class' parser (which doesn't exist) and promptly fails without getting to the next level of devices.

@philipstarkey What is the best way to hack around this? I've gone ahead with the simplest option I could think of which was adding the dummy intermediate device to the device registry with a very thin parser.

On a different note, I still need to get the runviewer parser to correctly account for wait and trigger delays. Quick advice or help would be appreciated.

dihm committed:
Nominal attempt to fix accounting for wait delays.

Tested in direct Clockline config without other things on the line

dihm committed:
Fix wait delay, make runviewer parser more accurate around waits.

dihm committed:
Fix bug when programming waits

dihm committed:
adding the sphinx docs for ultimate inclusion in labscript-devices

dihm committed:
Update docs about different triggering schemes.

dihm commented:
Pretty sure I have fixed the runviewer parser (as well as a bug in how waits were programmed).

I've also gone ahead and added the rst docs that will be ultimately used when adding to the `labscript-devices` repo.

Remaining tests:
- Test how sharing Clockline triggering with other PrawnDO and DAQ-like devices behaves (I'm worried ticks from other devices on the clockline will lead to extra programming instructions/collisions)
- Confirm that triggering with a normal device behaves as expected
- Make some interesting demonstrative pulse sequences

dihm commented:
Just realized that I never merged this. I believe I have done most of the remaining tests at some level. If there aren't any immediate objections, I'll go ahead and merge this in a couple days.

dihm committed:
Fix docs typos and remove no longer used import.
